File: M3Tools/Controls/Listeners/TemplateSelector.vb
Imports System.ComponentModel
Imports System.Windows.Forms

Public Class TemplateSelector
	' TODO: Create template management system for this later

	Private ReadOnly Property SelectedTemplate As Types.Template
		Get
			Return CType(cbx_TemplateSelection.SelectedItem, Types.Template)
		End Get
	End Property

	<[ReadOnly](True)>
	Public ReadOnly Property TemplateName As String
		Get
			Return cbx_TemplateSelection.SelectedText
		End Get
	End Property

	<[ReadOnly](True)>
	Public ReadOnly Property TemplateValue As String
		Get
			Return CStr(cbx_TemplateSelection.SelectedValue)
		End Get
	End Property
	Public ReadOnly Property TemplateSubject As String
		Get
			Return If(SelectedTemplate?.Subject, "")
		End Get
	End Property


	Sub Reload()
		cbx_TemplateSelection.SelectedIndex = -1
	End Sub

	Private Sub TemplateSelectionChanged(sender As Object, e As EventArgs) Handles cbx_TemplateSelection.SelectedIndexChanged
		If cbx_TemplateSelection.SelectedIndex < 0 Then
			cbx_TemplateSelection.Text = "Select Template..."
			Return
		End If

		'wb_TemplateDisplay.Navigate("about:blank")
		wb_TemplateDisplay.DocumentText = CStr(cbx_TemplateSelection.SelectedValue)
	End Sub

	Private Sub TemplateListUpdated(sender As Object, e As ListChangedEventArgs) Handles bsTemplates.ListChanged
		If Not e.ListChangedType = ListChangedType.ItemAdded Then
			Return
		End If

	End Sub

	Public Sub AddTemplate(template As Types.Template)
		bsTemplates.Add(template)
	End Sub

	Public Sub AddRange(templates As IList(Of Types.Template))
		For Each template In templates
			AddTemplate(template)
		Next
	End Sub
End Class

File: M3Tools/Controls/People/Listeners/TemplateSelector.cs
using System;
using System.ComponentModel;
using Microsoft.VisualBasic.CompilerServices;

namespace SPPBC.M3Tools
{
	/// <summary>
	/// 
	/// </summary>
    public partial class TemplateSelector
    {
        // TODO: Create template management system for this later

        private Types.Template SelectedTemplate
        {
            get
            {
                return (Types.Template)cbx_TemplateSelection.SelectedItem;
            }
        }

		/// <summary>
		/// The name of the currently selected template
		/// </summary>
        [ReadOnly(true)]
        public string TemplateName
        {
            get
            {
                return cbx_TemplateSelection.SelectedText;
            }
        }

		/// <summary>
		/// The value of the currently selected template
		/// </summary>
        [ReadOnly(true)]
        public string TemplateValue
        {
            get
            {
                return Conversions.ToString(cbx_TemplateSelection.SelectedValue);
            }
        }

		/// <summary>
		/// The subject for the currently selected template
		/// </summary>
        public string TemplateSubject
        {
            get
            {
                return (SelectedTemplate?.Subject) ?? "";
            }
        }

		/// <summary>
		/// 
		/// </summary>
        public TemplateSelector()
        {
            InitializeComponent();
        }


		/// <summary>
		/// Reloads the control
		/// </summary>
        public void Reload()
        {
            cbx_TemplateSelection.SelectedIndex = -1;
        }

        private void TemplateSelectionChanged(object sender, EventArgs e)
        {
            if (cbx_TemplateSelection.SelectedIndex < 0)
            {
                cbx_TemplateSelection.Text = "Select Template...";
				wb_TemplateDisplay.Navigate("about:blank");
                return;
            }

            wb_TemplateDisplay.DocumentText = TemplateValue;
        }

		/// <summary>
		/// Add a template to the list
		/// </summary>
		/// <param name="template"></param>
        public void AddTemplate(Types.Template template)
        {
            bsTemplates.Add(template);
        }

		/// <summary>
		/// Add a range of templates to the list
		/// </summary>
		/// <param name="templates"></param>
        public void AddRange(Types.TemplateList templates)
        {
            foreach (var template in templates)
                AddTemplate(template);
        }
    }
}
